ucan-discussion.md

```diff
@@ -0,0 +1,146 @@
+
+
+### Option 3 - A UUID
+
+Another option is to generate a UUID for each user to use as their bucket name.
+
+The advantages of this are that it’s:
+1. Short
+2. Globally unique across services
+3. Removes the link between key-pairs and the bucket name, allowing key-pairs to be rotated.
+
+This UUID could be synced between web3.storage and nft.storage to allow each user to have a single bucket identifier which is global. But unlike with option 2, the user doesn’t sync this value for us.
+
+
+Discussion
+----------
+
+It seems that the one advantage of using the DID as the bucket name is essentially this: it gets the user to do the work of setting a common value between nft.storage and web3.storage for us. As far as I (Adam) can see (but I might be wrong!) there is no other advantage to this over the UUID solution, it just gives us that syncing for free. The trade-off is the blurring of parameters and potential confusion that it may cause.
+
+Given that we already have the https://api.nft.storage/user/did endpoint to allow the user to specify their DID. Would it make sense to allow them to specify their bucket name in a similar way? This would allow a conceptual separation between buckets and authentication keys while still allowing them to have the same bucket name across all storage services.
+
+
+### Revocation / Recovery
+
+If a user loses their private key, then we want to provide a way for them to still control their bucket. This could be done in a couple of ways:
+1. Allow them to “rename” (i.e. move) their bucket.
+    - If so, we need to handle the mapping between old and new from a path perspective.
+2. Allow them to change their authorised keys.
+    - Control a list of keys in their account which are allowed to sign UCANs?
+    - Revoke UCANs which have already been issued? Would this be done automatically if the keys which signed the UCANs have been revoked?
+
+Revocation/recovery probably needs some more thought.
+
+
+Scenarios to consider
+---------------------
+
+### Recovery scenario
+
+This scenario applies if we are using the user’s public key/DID as their bucket name and we’re implementing an ability for a user to recover/reallocate their bucket if they lose their private key.
+
+1. User A uploads their public key; we don’t validate that they own it when registering (POST user/did).
+2. User B uploads User A’s public key as their own; again, we don’t validate that we own it when registering.
+3. User B then tells us that they’ve lost their private key, so they ask for their bucket to be mapped to a new public key and they provide us with a new public key.
+4. We transfer ownership of the bucket belonging to User B’s public key to the new key that they provided. We’ve now accidentally transferred control of User A’s bucket to User B.
```

Comment:
It seems that uploading a key here implies that user is given access to corresponding bucket, which it should not. User should be able to grant access to the bucket to specific key holder, that is user could add a key to specific namespace they control, if they gave it to a key they have to private key of they shared access to their stuff to some other keyholder as opposed to obtained access to some bucket they had no prior access to.